src/v/kafka/client/brokers.cc

```diff
@@ -95,4 +95,10 @@ ss::future<bool> brokers::empty() const {
     return ss::make_ready_future<bool>(_brokers.empty());
 }
 
+ss::future<> brokers::clear() {
+    _brokers.clear();
+    _next_broker = 0;
```

Comment:
Instead of trying to find all the state to clear, is it possible to do something like:

```c++
namespace pandaproxy::schema_registry

ss::future<> api::restart() {
    co_await stop();
    co_await start();
}

} // namespace pandaproxy::schema_registry

Reply:
What you say makes sense but does stopping and starting at the API level also reset the state for internal containers?
Looking at `schema_registry/api.cc` it looks like we'd run into the same issue reported in a previous review comment https://github.com/redpanda-data/redpanda/pull/7765#discussion_r1048865568

```cpp
ss::future<> api::stop() {
    co_await _service.stop();
    co_await _sequencer.stop();
    co_await _client.stop();
    if (_store) {
        co_await _store->stop();
    }
}
```

Reply:
Ah, turning the `_service`, `_sequencer`, and `_client` into smart pointers would do it.

Reply:
> What you say makes sense but does stopping and starting at the API level also reset the state for internal containers?

A quick glance suggests that `sharded_service<>::stop()` deletes the instances, I didn't see any problems then calling `sharded_service<>::start()` but I only glanced at the implementation.

As you suggest, an indirection via smart pointer should be possible.